HiAnalysis/HiOnia/plugins/HiOniaAnalyzer.cc

```diff
@@ -799,451 +864,696 @@ HiOniaAnalyzer::fillTreeMuon(const pat::Muon* muon, int iType, ULong64_t trigBit
-      Reco_QQ_ctau3D[Reco_QQ_size] = -100;
-      std::cout << "Warning: User Float ppdlPV3D was not found" << std::endl;
-    }
-    if (aJpsiCand->hasUserFloat("ppdlErrPV3D")) {
-      Reco_QQ_ctauErr3D[Reco_QQ_size] = 10.0*aJpsiCand->userFloat("ppdlErrPV3D");
-    } else {
-      Reco_QQ_ctau3D[Reco_QQ_size] = -100;
-      std::cout << "Warning: User Float ppdlErrPV3D was not found" << std::endl;
-    }
-  }
-  if (_isMC){
-    if (aJpsiCand->hasUserFloat("ppdlTrue")) {
-      Reco_QQ_ctauTrue[Reco_QQ_size] = 10.0*(float)aJpsiCand->userFloat("ppdlTrue");
-    } else {
-      Reco_QQ_ctauTrue[Reco_QQ_size] = -100;
-      std::cout << "Warning: User Float ppdlTrue was not found" << std::endl;
-    }
-    if (aJpsiCand->hasUserFloat("ppdlTrue3D")) {
-      Reco_QQ_ctauTrue3D[Reco_QQ_size] = 10.0*(float)aJpsiCand->userFloat("ppdlTrue3D");
-    } else {
-      Reco_QQ_ctauTrue3D[Reco_QQ_size] = -100;
-      std::cout << "Warning: User Float ppdlTrue3D was not found" << std::endl;
-    }
-  }
-  if (aJpsiCand->hasUserFloat("vProb")) {
-    Reco_QQ_VtxProb[Reco_QQ_size] = aJpsiCand->userFloat("vProb");
+  if (aBcCand==NULL){
+    std::cout<<"ERROR: 'aBcCand' pointer in fillTreeBc is NULL ! Return now"<<std::endl; return;
   } else {
-    Reco_QQ_VtxProb[Reco_QQ_size] = -1;
-    std::cout << "Warning: User Float vProb was not found" << std::endl;
-  }
-  if (aJpsiCand->hasUserFloat("DCA")) {
-    Reco_QQ_dca[Reco_QQ_size] = aJpsiCand->userFloat("DCA");
-  } else {
-    Reco_QQ_dca[Reco_QQ_size] = -99;
-    std::cout << "Warning: User Float DCA was not found" << std::endl;
-  }
-  if (aJpsiCand->hasUserFloat("MassErr")) {
-    Reco_QQ_MassErr[Reco_QQ_size] = aJpsiCand->userFloat("MassErr");
-  } else {
-    Reco_QQ_MassErr[Reco_QQ_size] = -99;
-    std::cout << "Warning: User Float MassErr was not found" << std::endl;
-  }
-  if (aJpsiCand->hasUserInt("Ntrk")) {
-    Reco_QQ_Ntrk[Reco_QQ_size] = aJpsiCand->userInt("Ntrk");
-  } else {
-    Reco_QQ_Ntrk[Reco_QQ_size] = -1;
-    std::cout << "Warning: User Int Ntrk was not found" << std::endl;
-  }
+    const pat::Muon* muon1 = dynamic_cast<const pat::Muon*>(aBcCand->daughter("muon1"));
+    const pat::Muon* muon2 = dynamic_cast<const pat::Muon*>(aBcCand->daughter("muon2"));
+    const pat::Muon* muon3 = dynamic_cast<const pat::Muon*>(aBcCand->daughter("muon3"));
+    
+    if (muon1==NULL || muon2==NULL || muon3==NULL){
+      std::cout<<"ERROR: 'muon1' or 'muon2' or 'muon3' pointer in fillTreeBc is NULL ! Return now"<<std::endl; return;
+    } else {
+      
+      int charge = muon1->charge() + muon2->charge()+ muon3->charge(); 
+      
+      TLorentzVector vMuon1 = lorentzMomentum(muon1->p4());
+      TLorentzVector vMuon2 = lorentzMomentum(muon2->p4());
+      TLorentzVector vMuon3 = lorentzMomentum(muon3->p4());
+
+      reco::TrackRef iTrack_mupl;
+      reco::TrackRef iTrack_mumi;
+      reco::TrackRef iTrack_muW;
+
+      int mu1_idx = IndexOfThisMuon(&vMuon1 , Reco_mu_4mom, false);
```

Comment:
I used a mapping of the value of muon Pt to the index of the muon.
For the key of the map (i.e. the Pt value), I had to use type "long int" (after appropriate 10^6 precision multiplication) to avoid rounding errors.